### PR TITLE
Resolve #12 Add retry mechanism

### LINKED ISSUE #12: Add retry mechanism to compiler for final-fidelity constraints
Application programmers should be able to specify a minimum "final-fidelity" for their EPR requests. This value indicates the minimum fidelity that all generated EPR qubits should have at the moment that the whole request has finished.

Application programmers should also be able to specify how many times the application should try to meet this minimum final-fidelity requirement. If the programmer indicates a try-count of more than 1, the application should automatically retry the EPR request until the final-fidelity requirement is met, or until the maximum number of tries has been reached.

This works involves changes to the SDK and to the compiler. This issue is for the changes in the **compiler** only. See [the other issue](https://github.com/QuTech-Delft/netqasm/issues/11) for the SDK changes.

## What needs to be done

The compiler should convert the `min_fidelity_all_at_end` value to a timeout value either using some hard-coded lookup table or some kind of heuristic function. This timeout value is the time within which the last EPR pair should have been generated. If the last pair is not generated within the timeout value, the compiler considers this as failing the fidelity requirement provided by the application programmer.

The compiler creates a loop in the generated NetQASM subroutine, where the maximum number of iterations is equal to max_tries provided through the SDK. In each iteration, the EPR request coming from the SDK is handled (using corresponding arrays and `create_epr`/`recv_epr` and `wait` instructions). Furthermore, the compiler inserts code such that:

- when the results for the last pair are available, the "Generation time" (see NetQASM EPR request/response interface) for the last pair is compared to the computed timeout value
- if the goodness time is below the threshold, a jump instruction will break out of the loop
- if the goodness time is above the threshold,
  -  if less than `max_tries` iterations have been made, a new iteration of the loop is started
  - if `max_tries` iterations have been reached, the compiler inserts an Abort instruction

### PR DESCRIPTION
**Pull request by bvdvecht**
Closes #12

### CODE REVIEW

File: netqasm/sdk/compiling.py [hunk 1]
```diff
@@ -266,8 +268,17 @@ def _move_carbon_electron(
...
-        qubit_id1 = self.get_reg_value(instr.reg1).value
+        try:
+            qubit_id0 = self.get_reg_value(instr.reg0).value
+            qubit_id1 = self.get_reg_value(instr.reg1).value
+        except KeyError:
+            # Register values are not known at compile time.
+            # We may assume that this is a MOV operation from the communication
+            # qubit to a memory qubit. (This is the only time a gate uses
+            # operands that are not known at compile time.)
+            assert isinstance(instr, vanilla.MovInstruction)
```
Comment (QFer) [inline netqasm/sdk/compiling.py:279]: shouldn't this assert not be done before the try?

File: netqasm/sdk/connection.py [hunk 2]
```diff
@@ -616,6 +616,31 @@ def loop_body(
         """
         self._builder.sdk_loop_body(body, stop, start, step, loop_register)
 
+    def while_true(self, max_iterations: int) -> ContextManager[SdkWhileTrueContext]:
+        """Create a context with code to be looped until the exit condition is met, or
+        the maximum number of tries has been reached.
+
+        The code inside the context is automatically looped (re-run).
+        At the end of each iteration, the exit_condition of the context object is
+        checked. If the condition holds, the loop exits. Otherwise the loop
```
Comment (QFer) [inline netqasm/sdk/connection.py:625]: As you describe it here it's not a while_true but a loop_until.

File: tests/test_builder.py [hunk 3]
```diff
@@ -579,6 +582,246 @@ def test_try():
...
+    with DebugConnection("Alice") as conn:
+
+        with conn.while_true(max_iterations=10) as loop:
+            q = Qubit(conn)
+            m = q.measure()
+            constraint = ValueAtMostConstraint(m, 42)
+            loop.set_exit_condition(constraint)
+
+        subroutine = conn.builder.subrt_pop_pending_subroutine()
+        print(subroutine)
```
Comment (QFer) [inline tests/test_builder.py:595]: I see a nice program with the while construction, but the test doesn't validate to true or false. When this test is run as a single test I get a 'TypeError: calling <function test_while_true at 0x0000021F556D7708> returned None, not a test'. Maybe check if the subroutine is defined correctly by checking all or parts of the commands list. Edit: I see a nice PreSubroutineInspector used in other tests.

**Review by QFer — CHANGES_REQUESTED** [review]

File: tests/test_builder.py (hunk 3 above)
Comment (bvdvecht) [reply tests/test_builder.py:595]: How are you running the test? When running `pytest tests/test_builder.py -k while_true` the test passes for me. In any case I also added a pattern match test.

File: netqasm/sdk/connection.py (hunk 2 above)
Comment (bvdvecht) [reply netqasm/sdk/connection.py:625]: It's indeed a better name. I replaced all `while_true` with `loop_until`.

File: netqasm/sdk/compiling.py (hunk 1 above)
Comment (bvdvecht) [reply netqasm/sdk/compiling.py:279]: No, the instruction can be something other than a Mov. However, when there is no key found, we only allow the situation where it is a Mov.

File: netqasm/sdk/connection.py (hunk 2 above)
Comment (QFer) [reply netqasm/sdk/connection.py:625]: I guess you should also replace WhileTrue in SdkWhileTrueContext then

File: tests/test_builder.py (hunk 3 above)
Comment (QFer) [reply tests/test_builder.py:595]: I run it as a single test.. the pattern match makes it a test with result

File: netqasm/sdk/connection.py (hunk 2 above)
Comment (bvdvecht) [reply netqasm/sdk/connection.py:625]: Done

**Review by QFer — APPROVED** [review]

**Pull request merged**